tests/functional/verification/test_adt_1.py

```diff
@@ -0,0 +1,54 @@
+from nagini_contracts.adt import ADT
+from nagini_contracts.contracts import *
+
+from typing import (
+    cast,
+    List,
+    NamedTuple,
+)
+
+class LinkedList(ADT):
+    pass
+
+class Node(LinkedList, NamedTuple('Node', [('elem', int), ('next', LinkedList)])):
+    pass
+
+class Null(LinkedList, NamedTuple('Null', [])):
+    pass
+
+def common_use_of_ADTs()-> None:
+    l_1 = Null()
+
+    l_2 = Node(2, Null())
+    x = l_2.elem
+    assert l_2.elem == x
+    Assert(l_2.elem == 2)
+
+    l_3 = Node(3, Node(4, Null()))
+    assert cast(Node, l_3.next).elem == 4
+
+def preppend(l: LinkedList, elem: int) -> LinkedList:
+    Ensures(Result() == Node(elem, l))
+    return Node(elem, l)
+
+def constructor_totality_1(l: LinkedList) -> None:
+    if type(l) is Node:
```

Comment:
I think we could use some more tests, e.g.:
- Same as this one, but with  ``isinstance`` instead of ``type``
- Using something like ``l.elem`` within a contract
- Passing a LinkedList to a function that expects something of type object (and then casting it back to a LinkedList in the body of that function)
- Using an ADT in a global variable, a predicate, and a pure function
- Having two ADTs, and a field of the second one whose type is the first ADT
- ADT expressions that are used in complex expressions. E.g. if we have a normal class A with an int field g, and an ADT B with a constructor C with a field f of type A, an expression like C(A()).f.g